File: admin/src/main/scala/com/neu/model/Admission.scala
package com.neu.model

case class AdmRuleSubCriterion(
  name: String,
  op: String,
  value: String
)

case class AdmRuleCriterion(
  sub_criteria: Option[Array[AdmRuleSubCriterion]],
  name: String,
  op: Option[String],
  value: Option[String],
  path: Option[String],
  template_kind: Option[String],
  `type`: Option[String],
  value_type: Option[String]
)

case class AdmRule(
  id: Int,
  category: String,
  comment: Option[String],
  criteria: Option[Array[AdmRuleCriterion]],
  disable: Option[Boolean],
  critical: Option[Boolean] = None,
  cfg_type: Option[String] = Some("user_created"),
  rule_type: Option[String]
)

case class AdmRuleConfig(
  config: AdmRule
)

case class AdmState(
  enable: Boolean,
  mode: String,
  default_action: String,
  adm_client_mode: String,
  adm_client_mode_options: Map[String, String]
)

case class AdmConfig(
  state: AdmState
)

case class AdmExport(
  ids: Array[Int],
  export_config: Boolean
)

File: admin/webapp/websrc/app/common/types/admission/admission.ts
export class AdmRuleSubCriterion {
  name: string = "";
  op: string = "";
  value: string = "";
  unit?: string;
}

export interface AdmRuleCriterion {
  sub_criteria?: Array<AdmRuleSubCriterion>;
  name: string;
  op?: string;
  value?: string;
}

export interface AdmissionRule {
  id?: number;
  category?: string;
  comment?: string;
  criteria?: Array<AdmRuleCriterion>;
  disable?: boolean;
  critical?: boolean;
  cfg_type?: string;
  rule_type?: string;
}

export interface AdmissionRules {
  rules: Array<AdmissionRule>;
}

interface AdmissionState {
  enable?: boolean;
  mode?: string;
  default_action?: string;
  adm_client_mode?: string;
  adm_svc_type?: string;
  failure_policy?: string;
  adm_client_mode_options?: {
    [key: string]: string[];
  };
  ctrl_states?: {
    [key: string]: boolean[];
  };
  cfg_type: string;
}

export interface AdmissionStateRec {
  k8s_env: boolean;
  state?: AdmissionState;
}

export interface MatchedRule {
  id: number;
  disabled: boolean;
  container_image: string; 
  type: string;
  mode: string;
  rule_details: string;
  rule_cfg_type: string;
}

export interface AdmissionTestResult {
  allowed: Boolean;
  index: number;
  kind: string;
  message: string;
  name: string;
  matched_rules: MatchedRule[] | null;
}

export interface AdmissionConfigurationAssessment {
  props_unavailable: Array<string>;
  global_mode: string;
  results: Array<AdmissionTestResult>;
}
